File: 66-plus-one/66-plus-one.py
class Solution(object):
    def plusOne(self, digits):
        """
        :type digits: List[int]
        :rtype: List[int]
        """
        index = 0 # End of the array
        mustIncrement = True
        
        while mustIncrement: # Handle digit increments for all indices.
            index -= 1
            
            if index < -len(digits): # Must add a column and shift the array.
                digits.insert(0, 1) # Add the new most significsant digit.
                mustIncrement = False
            else: # Simply increment index digit and move on.
                digits[index] = (digits[index] + 1) % 10
                mustIncrement = digits[index] == 0
            
        return digits

File: 66-plus-one/66-plus-one.java
class Solution {
    
    public int[] plusOne(int[] digits) {
        
        // Solution: O(n) time, O(1) additional space.
        
        int n = digits.length;
        
        for (int i = n - 1; i >= 0; i--) {
            
            if (digits[i] < 9) {
                digits[i] += 1;
                return digits;
            }
            
            digits[i] = 0; // When digit is 9, have to set digit to 0 and carry.
            
        }
        
        // Handle most significant digit carries that need the array to be extended.
        digits = new int[n + 1];
        digits[0] = 1;
            
        return digits;
        
    }
    
}